### DISCUSSION TITLE
Dependency Injection and Field Validation

### DISCUSSION THREAD

**Discussion by ouyangqr** [fastapi/fastapi#12939] [Questions] [answered]
### Privileged issue

- [X] I'm @tiangolo or he asked me directly to create an issue here.

### Issue Content

code
```
def get_db():
    db = SessionLocal2()
    try:
        yield db
    except Exception as e:
        logger.exception(e)
        db.rollback() 
        raise e
    finally:
        db.close()
class NurseNewSchema(BaseModel):
    family_user_id: str
    time_day: int
    nurse_master: str
    nurse_remarks: str | None
    nurse_title: str

async def post_new_nurse(item: NurseNewSchema, db=Depends(get_db)):
    pass
```
When nurse_master is not filled in, data verification fails, but the interface executes to get_db, and when yielding db, it reports data verification failure.

2024-11-06 09:58:29.965 | ERROR    | core:get_db:59 - [{'loc': ('body', 'nurse_master'), 'msg': 'field required', 'type': 'value_error.missing'}]
Traceback (most recent call last):

  File "/usr/lib/python3.10/threading.py", line 973, in _bootstrap
    self._bootstrap_inner()
    │    └ <function Thread._bootstrap_inner at 0x7fd6edf39090>
    └ <WorkerThread(AnyIO worker thread, started 140559949760064)>
  File "/usr/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
    │    └ <function WorkerThread.run at 0x7fd6c1500430>
    └ <WorkerThread(AnyIO worker thread, started 140559949760064)>
  File "/home/ouyang/.env/paper/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 807, in run
    result = context.run(func, *args)
             │       │   │      └ (<class 'fastapi.exceptions.RequestValidationError'>, RequestValidationError([{'loc': ('body', 'nurse_master'), 'msg': 'field...
             │       │   └ <bound method _GeneratorContextManager.__exit__ of <contextlib._GeneratorContextManager object at 0x7fd6c154d930>>
             │       └ <method 'run' of '_contextvars.Context' objects>
             └ <_contextvars.Context object at 0x7fd6d9e22700>
  File "/usr/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
    │    │   │     │    │      └ None
    │    │   │     │    └ RequestValidationError([{'loc': ('body', 'nurse_master'), 'msg': 'field required', 'type': 'value_error.missing'}])
    │    │   │     └ <class 'fastapi.exceptions.RequestValidationError'>
    │    │   └ <method 'throw' of 'generator' objects>
    │    └ <generator object get_db at 0x7fd6c1537b50>
    └ <contextlib._GeneratorContextManager object at 0x7fd6c154d930>

> File "/home/ouyang/Desktop/poc/core.py", line 55, in get_db
    yield db
          └ <sqlalchemy.orm.session.Session object at 0x7fd6c154de10>
fastapi.exceptions.RequestValidationError: [{'loc': ('body', 'nurse_master'), 'msg': 'field required', 'type': 'value_error.missing'}]

**Comment by prajwalmj07** [comment]
The error is happening because FastAPI is trying to yield the db session in get_db() before it handles the validation error. By ensuring that the validation happens before calling the dependency (get_db), the session will only be created if the input is valid. This prevents unnecessary database connections in case of invalid input.  

1)Refactor get_db() as a Context Manager: This keeps the session logic clear and prevents unnecessary session creation before validation.
2)Ensure Validation Happens First: FastAPI will validate item (which is the body input) before get_db is even called. So if there’s any validation failure, it will happen before the DB session is created.
3)Error Handling in get_db(): If an exception is raised (such as a validation error), get_db() will log and handle it without impacting the database session.

**Comment by dolfinus** [comment]
> This prevents unnecessary database connections in case of invalid input.

In production applications it is common to create SQLAlchemy's Engine, and then use it so create sessions. Engine is a connection pool, so it doesn't really create new session each time, but reuses already opened session from the pool.

**Comment by YuriiMotov** [comment, marked as answer]
There is no way to specify the order in which dependencies will be solved in FastAPI. 

But it's not a problem with SQLAlchemy connections - as you've mentioned, using connection pool real connection will only be acquired when you read or write data from\to DB. So, this dependency call is cheap.

To avoid printing unnecessary errors in log: instead of handling all exceptions in `get_db` (`except Exception as e:`), you should specify classes of exceptions you really need to handle.
